src/pages/library/SearchAndFilterComponent.ts

```diff
@@ -291,12 +291,20 @@ const FilterSectionComponent = <ListItem>({ section, fullList, selectedSections,
   const itemsFilteredByOtherSections = getMatchingDataForSectionList(selectedSectionsWithoutSection, fullList)
   // We want to filter out values for which there are no entries. This is only really important for values that are not derived from
   // the list data itself, like dataset access levels. However, because we don't differentiate between types of filters, we need to test all of them
-  const valuesToShow = _.filter(sectionEntry => _.size(listItemsMatchForSectionEntry(sectionEntry, section.matchBy, fullList)) > 0, section.values)
+  const labelsWithEntries = _.filter(sectionEntry => _.size(listItemsMatchForSectionEntry(sectionEntry, section.matchBy, fullList)) > 0, section.values)
+  const selectedLabels = _.filter(sectionEntry => sectionEntrySelected(section, sectionEntry, selectedSections), section.values)
+  // The shown labels are the first N labels that have any catalog entries, with any selected labels added at the end,
+  // if they aren't one of the first N labels.
+  const shownLabels = _.flow(
+    _.intersection(labelsWithEntries),
+    _.concat(_.take(numLabelsToRender, labelsWithEntries)),
+    _.uniq
+  )(selectedLabels)
 
   return h(Fragment, [
-    _.map(sectionEntry => {
+    _.map((sectionEntry : string) => {
       const numMatches = listItemsMatchForSectionEntry(sectionEntry, section.matchBy, itemsFilteredByOtherSections).length
-      const sectionEntryChecked = sectionEntrySelected(section, sectionEntry, selectedSections)
+      const sectionEntryChecked = selectedLabels.includes(sectionEntry)
```

Comment:
nit: For consistency (and other reasons) I believe we still use lodash functions over the native functions.

For reference this can be found in the Lodash section of the [contributor guide](https://github.com/DataBiosphere/terra-ui/wiki/Contributor-Guide)

```suggestion
      const sectionEntryChecked = _.includes(sectionEntry, selectedLabels)
```